File: packages/shared-ts/src/idempotency.ts
// Idempotency-key store (instructions.md §21, Principle #8).
// Wire-compatible contract with packages/shared-py/olma_shared/idempotency.py.
// Prod backs this with Redis (24h TTL, §16.2); this is the in-memory dev impl.

export interface IdempotencyStore {
  remember(key: string, value: unknown, ttlSeconds?: number): boolean;
  get(key: string): unknown | null;
  seen(key: string): boolean;
}

export class InMemoryStore implements IdempotencyStore {
  private data = new Map<string, { exp: number; value: unknown }>();

  private purge(now: number): void {
    for (const [k, v] of this.data) if (v.exp <= now) this.data.delete(k);
  }

  /** Store value if absent. Returns true if newly stored, false if it existed. */
  remember(key: string, value: unknown, ttlSeconds = 86400): boolean {
    const now = Date.now() / 1000;
    this.purge(now);
    if (this.data.has(key)) return false;
    this.data.set(key, { exp: now + ttlSeconds, value });
    return true;
  }

  get(key: string): unknown | null {
    const now = Date.now() / 1000;
    const entry = this.data.get(key);
    if (!entry) return null;
    if (entry.exp <= now) {
      this.data.delete(key);
      return null;
    }
    return entry.value;
  }

  seen(key: string): boolean {
    return this.get(key) !== null;
  }

  clear(): void {
    this.data.clear();
  }
}

File: packages/shared-py/olma_shared/idempotency.py
"""Idempotency-key store (instructions.md §21, Principle #8).

Every write carries an Idempotency-Key; a retry must never produce a duplicate.
Prod backs this with Redis (24h TTL, §16.2); this module defines the interface
and a process-local implementation for dev/tests.
"""

from __future__ import annotations

import time
from typing import Any, Protocol


class IdempotencyStore(Protocol):
    def remember(self, key: str, value: Any, ttl: float = 86400) -> bool: ...
    def get(self, key: str) -> Any | None: ...
    def seen(self, key: str) -> bool: ...


class InMemoryStore:
    """Process-local store with TTL. Not shared across replicas (dev only)."""

    def __init__(self) -> None:
        self._data: dict[str, tuple[float, Any]] = {}

    def _purge(self, now: float) -> None:
        expired = [k for k, (exp, _) in self._data.items() if exp <= now]
        for k in expired:
            del self._data[k]

    def remember(self, key: str, value: Any, ttl: float = 86400) -> bool:
        """Store value for key if absent. Returns True if newly stored, False if it existed."""
        now = time.time()
        self._purge(now)
        if key in self._data:
            return False
        self._data[key] = (now + ttl, value)
        return True

    def get(self, key: str) -> Any | None:
        now = time.time()
        entry = self._data.get(key)
        if entry is None:
            return None
        exp, value = entry
        if exp <= now:
            del self._data[key]
            return None
        return value

    def seen(self, key: str) -> bool:
        return self.get(key) is not None

    def clear(self) -> None:
        self._data.clear()
